New approach to location-safe HQ reports

esoergel opened the pull request:
https://dimagi-dev.atlassian.net/browse/SC-759

The previous approach worked by updating a global when the report class was first loaded, but that wasn't reliable.  Not all report classes are guaranteed to have been loaded.  The major downside of this new approach is that the safety flag is now inherited.  The same behavior exists for Django CBVs though, so it seems a reasonable tradeoff.

I would be interested to hear alternative ideas to get the best of both worlds, however.  What I really want is a way to identify classes that were directly decorated, not their children (or worse, ancestors!).  One idea I had was rather than setting `is_location_safe` to `True`, set it to the name of the class, then compare that.  Still seems messy, but it might work better.  Any ideas?

This also adds tests, which makes it a lot easier to iterate.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
Holding off on merge at the moment, as I want to audit the existing reports for the inheritance problem (I know at least the CaseListReport has to deal with this).

##### PRODUCT DESCRIPTION
Fixes a bug where the Case List Explorer (and possibly others?)  Would sometimes incorrectly give 403 responses to location-restricted users.

esoergel committed:
add basic location safety flag tests

esoergel committed:
Clarify current behavior regarding inheritance

esoergel committed:
Add tests for HQ reports

Man this system is convoluted

esoergel committed:
Convert a couple dispatcher methods to classmethods

esoergel committed:
Test conditionally safe views

esoergel committed:
Test conditionally safe reports

esoergel committed:
Switch flagging approach for HQ reports

The previous approach worked by updating a global when the report class
was first loaded, but that wasn't reliable.  Not all report classes are
guaranteed to have been loaded.  The major downside of this new approach
is that the safety flag is now inherited.  The same behavior exists for
Django CBVs though, so it seems a reasonable tradeoff.

esoergel committed:
lint

millerdev committed:
Do not inherit location safety from parent class

millerdev reviewed (approved):
I added https://github.com/dimagi/commcare-hq/commit/5ac630b3671d7b1c52d84c1cdad92b567cb3cb9d on a fork of your branch. Feel free to cherry-pick it to this PR if you like it.

esoergel committed:
This now only uses attrs set on the class itself

esoergel committed:
Apply get_own_property logic to conditional views

millerdev commented on corehq/apps/locations/permissions.py:
```diff
@@ -239,27 +235,31 @@ def is_location_safe(view_fn, request, view_args, view_kwargs):
-        return view_fn._conditionally_location_safe_function(view_fn, request, *view_args, **view_kwargs)
+    conditional_fn = _get_own_property(view_fn, '_conditionally_location_safe_function')
+    if conditional_fn:
+        return conditional_fn(view_fn, request, *view_args, **view_kwargs)
```
Does it matter that `view_fn` may have been reassigned to a class (`view_fn = view_fn.view_class`) above?

esoergel replied:
That's by design - my goal here is to store `is_location_safe` or `_conditionally_location_safe_function` directly on the thing that handles the request, whether a function view, class view, or HQ report.  That's whatever we developers would consider the main model for handling a request.  Then the first part of this function can normalize to that, and the boolean or conditional check can be evaluated in one place.

I think `view_fn` is probably no longer the best name for that though - got any ideas?

Here's refactor of the above if it makes the meaning clearer:

```python
def _view_obj_is_safe(obj, request, *view_args, **view_kwargs):
    if _get_own_property(obj, 'is_location_safe'):
        return True
    conditional_fn = _get_own_property(obj, '_conditionally_location_safe_function')
    if conditional_fn:
        return conditional_fn(obj, request, *view_args, **view_kwargs)


def is_location_safe(view_fn, request, view_args, view_kwargs):
    if 'resource_name' in view_kwargs:  # Tastypie
        return view_kwargs['resource_name'] in LOCATION_SAFE_TASTYPIE_RESOURCES

    if getattr(view_fn, 'is_hq_report', False):  # HQ report
        dispatcher = view_fn.view_class
        report_class = dispatcher.get_report(view_kwargs['domain'], view_kwargs['report_slug'])
        return _view_obj_is_safe(report_class, request, *view_args, **view_kwargs)

    if hasattr(view_fn, "view_class"):  # Django view
        return _view_obj_is_safe(view_fn.view_class, request, *view_args, **view_kwargs)

    return _view_obj_is_safe(view_fn, request, *view_args, **view_kwargs)
```

esoergel replied:
I should also note that there's a tight coupling between this and the decorator `location_safe`.  The approach I'm settling on is to make the decorator simpler - where it just sets a property on the object, and have this handle the complexity of the possible different types of objects.  If you look higher in the diff, you can see that the decorator used to set the property differently depending on what was being decorated.

millerdev replied:
:+1: sounds good

millerdev reviewed (commented):
Approved if the one question is not an issue.

esoergel committed:
These decorators are meant to be applied at the class level

esoergel committed:
Pull out safety check

esoergel committed:
Allow location safety to be inherited

This was the previous behavior for django views, and apparently ICDS relies on
that behavior quite extensively, so I'm not comfortable making the change.  This
switches reports to also inherit safety, but the CaseListReport is the only
location-safe report actually inherited, and that looks okay to do.

esoergel commented:
@millerdev apologies for the churn on this PR.  I just added a commit to revert the change to inheritance

> This was the previous behavior for django views, and apparently ICDS relies on
that behavior quite extensively, so I'm not comfortable making the change.  This
switches reports to also inherit safety, but the CaseListReport is the only
location-safe report actually inherited, and that looks okay to do.

millerdev reviewed (approved):
>  I just added a commit to revert the change to inheritance

Oh, that's kinda sad. I guess it would be too confusing to have a different behavior for ICDS than for vanilla HQ. It does seem better to not inherit safety, and ICDS reports (even though they depend on that) are probably not following the best practice there.

esoergel commented:
> Oh, that's kinda sad. I guess it would be too confusing to have a different behavior for ICDS than for vanilla HQ. It does seem better to not inherit safety, and ICDS reports (even though they depend on that) are probably not following the best practice there.

Totally agree - I'd be up for the change too, but don't want the fix here to be blocked by it.  I think it'd entail chasing down all descendants of decorated classes, and applying the decorator to them, which I imagine would be tedious, but not complicated.
